### ISSUE TITLE
Optimize parSequence

### ISSUE THREAD

**Issue by neko-kai** [zio/zio#2371]
As per @Avasil's [benchmarks](https://gist.github.com/pk044/181b32f9af380c9230d8512d14233c2a) (https://github.com/zio/zio/pull/2290#event-2821787971) `ZIO.parSequence` has a long way to go to get good performance. Any contributions to optimize it or related methods are welcome!

**Comment by vasilmkd** [comment]
Possibly a duplicate of #1182, but it needs more visibility. Either way, linking to the previous issue.

**Comment by horned-sphere** [comment]
I've made attempt at this in [this PR](https://github.com/zio/zio/pull/2492). The implementation is fairly ugly but it does work and is faster (I've attached the benchmark comparison). Any feedback on it would be greatly appreciated. 

If you think it has any merit, the change in implementation has broken an assumption of a test case in the stack tracer which I'd like to get some advice on before trying to fix it.
